engine/gamesys/src/gamesys/resources/res_sound_data.cpp

```diff
@@ -13,11 +13,68 @@
 // specific language governing permissions and limitations under the License.
 
 #include <string.h>
+#include <dlib/math.h>
+#include <dlib/mutex.h>
 #include <sound/sound.h>
+#include <resource/resource.h>
 #include "res_sound_data.h"
 
+#include <stdio.h> // printf
+
 namespace dmGameSystem
 {
+    static const uint32_t SOUNDDATA_DEFAULT_CACHE_SIZE = 2 * 1024*1024;
+    static const uint32_t SOUNDDATA_DEFAULT_CHUNK_SIZE = 16 * 1024;
+
+    struct SoundDataContext
+    {
+        dmMutex::HMutex m_Mutex;
+
+        // The cache size for all streaming sounds
+        uint32_t m_CacheSize;
+        // The chunk size when streaming in more data
+        uint32_t m_ChunkSize;
+    };
+
+    struct SoundDataChunk
+    {
+        uint8_t* m_Data;
+        uint32_t m_Size;
+        uint32_t m_Offset; // Offset into the file
+    };
+
+    struct SoundDataResource
+    {
+        dmSound::HSoundData       m_SoundData;
+        dmSound::SoundDataType    m_Type;
+
+        // For the streaming
+        SoundDataContext*         m_Context; // the global context for .oggc and .wavc respectively
+        dmResource::HFactory      m_Factory;
+        dmArray<SoundDataChunk*>  m_Chunks;
+        const char*               m_Path;
+        uint32_t                  m_FileSize:31;
+        uint32_t                  m_RequestInFlight:1;        // Have we already requested the next chunk?
+    };
+
+    static SoundDataContext* g_SoundDataContext = 0;
+
+    dmSound::HSoundData ResSoundDataGetSoundData(SoundDataResource* resource)
+    {
+        return resource->m_SoundData;
+    }
+
+    void ResSoundDataSetStreamingCacheSize(uint32_t cache_size)
+    {
+        g_SoundDataContext->m_CacheSize = cache_size;
+    }
+
+    // set the size of each streaming chunk size
+    void ResSoundDataSetStreamingChunkSize(uint32_t chunk_size)
+    {
+        g_SoundDataContext->m_ChunkSize = chunk_size;
+    }
```

Comment:
Wip. For now, we don't have config file access in this file, so the comp_sound.cpp will have to provide these values for now.